### ISSUE TITLE
Plugin system for complex queries

### ISSUE THREAD

**Issue by pushrbx** [Squidex/squidex#123]
https://github.com/pushrbx/squidex/commit/eb8ee8768f3860dba63720d52905e0373ac09190  (proof of concept)

I'd like to propose a plugin system for complex queries. Queries like "I'd like to have those articles which have these tags." This solution tries to solve the situation where you have to write multiple queries in your client to get these relationships, knowing that this kind of logic should belong on the server side.

With this squidex can load [MEF2 conmponents](https://www.nuget.org/packages/Microsoft.Composition/) from a folder.. Each component has a name, and attributes which describe what app and schema it's for. Every component has to provide a single operation. 
These operations then used by a controller which handles the following route:
```https://cloud.squidex.io/api/content/<app_name>/<schema_name>/operation/<custom_operation_name>```
The controller searches for the operation by going through the MEF components, and executes the exposed operation and passes a `ContentQueryService` object to the `ExecuteAsync` method of the operation. The operation then can do whatever with the query service, and it should return a result set.

So basically there would be a folder with bunch of dlls, each would represent a "custom query operation".  These operations then would be callable through the route above  and they would be tied to a single app and schema pair.

Interesting thing is that these "operations" can be used as views as well.

### TODOs:
- Improve the idea to make it more user friendly. Currently only those who self host squidex could benefit from this.
- Add graphql support
- Add UI to allow upload of plugin dlls. (how would one solve the security implications of this?)
- Refine the design

### Links
- https://dotnetthoughts.net/using-mef-in-dotnet-core/
- https://weblogs.asp.net/ricardoperes/using-mef-in-net-core

What do you think? 😄

**Comment by SebastianStehle** [comment]
I am not sure, if MEF is the best idea. The problem is Security and unloading them.

I would prefer to upgrade the scripting capabilities:
https://medium.com/squidex/updates-from-squidex-e3e6550d8cd1

In theory you just have to expose the functions to the script engine and you can then do whatever you want.

**Comment by SebastianStehle** [comment]
I like the idea, you could event materialize the queries then, you just need surrogate keys :)

**Comment by pushrbx** [comment]
I thought about (csx) scripts too, but I thought it would introduce too much stuff, so I rather aimed for MEF. But with scripting we could create something like Azure Functions, but more limited to a context defined by us.

**Comment by SebastianStehle** [comment]
The main challenge was to create fast wrappers over ContentData, but scripting with Jint is relatively easy. Btw: It is also used by RavenDb ;)

**Comment by pushrbx** [comment]
I could live that Jint. How about if I try to improve the existing Jint stuff by combining it with part of the above idea?

**Comment by SebastianStehle** [comment]
Something like that:

	// You have the following functions

	findByIds(schema, ids, callback);
	findByQuery(schema, query, callback);

	function QuerNodes(term) {
		findByQuery('terms', `data/term/iv eq ${term}`, function (terms) {
			if (terms.items.length === 1) {
				findByIds('nodes', terms.items[0].data.nodes.iv, function (nodes) {
					done(nodes);
				}
			}
			
			done([]);
		});
	}

I am open for other ideas, but I could not find a way to protect the system from CSharpScript. You can even crash the system with infinite loops and so on.

**Comment by pushrbx** [comment]
https://github.com/sebastienros/jint/issues/14
There is an option in Jint to prevent infinite loops.

> You actually define the maximum number of statements an Execute() call can run. If reached, it will throw a StatementsCountOverflowException. If the user's code does too many recursions you will get a StackoverflowException sent by the CLR itself. These are the two exceptions you should catch to provide specific error feedback, on top of JavaScriptException which will handle all standard ECMAScript ones.

**Comment by SebastianStehle** [comment]
Yes, in Jint there are a lot of options, but not in C# ;) ... I think you cannot catch a StackoverflowException.

**Comment by pushrbx** [comment]
That's true, you can't catch StackoverlowException in .net core. 
I would conclude at this point that Jint is fine I think, CsharpScript is not necessary, and it is not possible to prevent infinite loops, and other programming errors, or at least in this case it is out of scope. Max we could implement something which would give a proper error message. The only prevention can be done I think, that only the admin can add these scripts to schemas, and in case of cloud.squidex.io it should be a paid service or something else. Or shouldn't be available at all, rather available in a different pricing package where you provide a dedicated hosting for the customer. This way you can make sure that only "professionals" are creating scripts for the schemas.
Please assign me to this issue, and I try to put together something this weekend using Jint.
So my goal is:
- Allow user to add a script for a schema
- The script should be available through a route. e.g.: `api/content/<app_name>/<schema_name>/operation/<script_name>`
- The script should not be allowed to do mutations for now.
- The script should only accept GET requests.
- The script should be able to accept arguments from the url. (query string)

**Comment by SebastianStehle** [comment]
Sounds cool. There is already a script UI, would be cool if we can do something like in my example: use function names as operation Names.

**Comment by SebastianStehle** [comment]
Hi, have you started something?

**Comment by pushrbx** [comment]
Yes, I've started creating a new controller, and some extra stuff. I'll create a pull request when it's ready.

**Comment by SebastianStehle** [comment]
I also played around for 2 hours: https://github.com/Squidex/squidex/tree/feature-complex-query 

But it is not tested

**Comment by pushrbx** [comment]
Oh you did exactly what I wanted in the scripting part. 😄 But I've put it in the wrong namespace I guess. I'm going to reuse your attempt if you don't mind. 😄

**Comment by SebastianStehle** [comment]
My idea was to make a script file with functions:

	function mySpecialQuery(param) {
		var id = somethingWithParam(param);
		
		findContent('labels', id, function (result) {
			cb(result);
		});
         }

And then just use the function as controller argument:

	api/content/<app_name>/<schema_name>/queries/mySpecialQuery('123')

Then it is super easy to pass in complex arguments, even Objects and arrays. Of course you could als hack scripts into it, but it is only a query, there is no almost no potential to hack something (I hope :D)

**Comment by SebastianStehle** [comment]
Unfortunately I have no idea how to make it work with GraphQL because you need a schema for that. With this approach you can just return the count if you want.

**Comment by pushrbx** [comment]
My idea was to have a folder with script files.
`api/content/<app_name>/<schema_name>/queries/{scriptName}`
And it would run the script file from:
`Scripts/{app_name}/{schema_name}/{scriptName}.js`

And later improve this to have these scripts in Db, and create a UI where you can add them.
It seems we almost though about the same thing.

GraphQL: I think the solution will be something runtime generated with reflection. I'm not yet familiar with GraphQL in dotnet.

**Comment by SebastianStehle** [comment]
@GraphQL: Yes, but Javascript does not have reflection ;)

**Comment by SebastianStehle** [comment]
I would validate first if the general approach works, e.g. by making some integration tests as a did for GraphQL. If this works we can think about the details.

**Comment by pushrbx** [comment]
By reflection I meant to create a C# function at runtime which would forward the call to Jint by loading the script and passing it to the Engine. So I imagine something which would check the graphql query, if it contains a name with some prefix which would identify the name that it's going to be a javascript function, then it would create the function in runtime, add it to the GraphQL schema, then pass the query to graphql. 😄 

Btw I did my changes:
https://github.com/pushrbx/squidex/commit/ffaaeefadfe9998a101062ad91254eb10df29e64

**Comment by pushrbx** [comment]
Added pull request #130 just to see what our changes are.

**Comment by SebastianStehle** [comment]
## About Custom Queries

I think we have two options:

1. Queries are created for a specific schema. Then you know the structure and can also enforce that you are not returning anything unexpected, e.g. a number instead of a list of content items. Because I think that graphql-dotnet cannot handle that. For this use case we already have a UI. https://medium.com/squidex/updates-from-squidex-e3e6550d8cd1. We just need another tab for it and some fields in some objects. One of the challenges is to detect the functions in the script and also make it available to the GraphQL schema. You also have to make them available to the swagger specification.

2. Queries are not coupled to a schema. You can just do anything you want, as long as it is json. You cannot use them for GraphQL, because they are typed. But I guess some of use cases would require a map-reduce architecture and I don't want to write my own database :D

So I would go for option 1, my code also supports option 2 so far, so we have to move it to the right direction.

## About Plugins

I also saw your pull request for the plugin system. For me it is a little bit out of scope or I would not mix those too use cases together. Because a pugin system has other problems:

1. You need a contract / extension points, perhaps with versioning also.
2. You have to think about how to load them (and unload them).
3. Plugin systems are complicated for multi-node deployments.

I am not sure if I simple plugin system is very useful. You can just fork Squidex and make your own extension. This is what some users do. ;)

**Comment by pushrbx** [comment]
> You can just fork Squidex and make your own extension. This is what some users do


I did that, but it's just pain in the butt to keep on track with mainline branch, so I thought "i only need complex queries, and im fine, let's propose something which would solve it." ofc the plugin system came in my mind first. So basically I wanted to have complex queries through plugins, because that seemed easy to do. Now we are at that stage, that we are not really talking about a plugin system (that is my impression 😄 ) rather about having the complex query feature in squidex.

> So I would go for option 1

Me too.

> One of the challenges is to detect the functions in the script and also make it available to the GraphQL schema.

**Making them available**: My rough idea for this so far is to have something dynamic.
When the user adds the javascript, the system should run the script on submission to get the return types, remove the cached graphql schema, and build a type which inherits from `ObjectGraphType`  at runtime. `ContentQueryGraphType` would need an extra dependency on a list of scripts or it should expose a method to allow addition of extra fields. `GraphQLModel` class would need to expose a method too, to allow the addition of functions from scripts. And somewhere there along the lines there would a service which would execute the script, get the return type, and build a type which implements `IGraphType` for the `ResolvedType` property of `FieldType`.
Argh, bit difficult to phrase it, but I have a clue now by writing it down.

**Detecting functions** can happen on startup, I think they can be registered in the db. I can imagine this by a set of inputs around the textbox where the user types in the javascript, which would allow the user to set the name of the function and the parameters. That way it would have some sort of `model` as a starting point.

**Comment by SebastianStehle** [comment]
We could also start with something like an extension point for queries and then write the javascript system on top of it. I am fine with.

1. The plugin loading can be done with assembly scanning. Autofac already supports it: http://docs.autofac.org/en/latest/register/scanning.html. Should be 10 lines of code or so :D

2. The plugin extension must be more dynamic e.g.

	class QueryArgument 
	{
		public string Name { get; }
		public string Description { get; }
		
		public ArgumentType Type { get; }
	}
	
	enum QueryArgumentType
	{
		String,
		Boolean,
		Number
	}
	
	interface IQuery
	{
		IReadOnlyCollection<QueryArgument> Arguments { get; }
		
		string Name { get; }
		
		Task<(...)> Execute(QueryContext context, object[] arguments);
		
	}
	
	interface IQueryProvider
	{
		IEnumerable<IQuery> GetQueries(IAppEntity app, ISchemaEntity schema);
	}

**Comment by SebastianStehle** [comment]
On top of that you can either write your custom queries with C# or I/we can write the javascript system.

**Comment by pushrbx** [comment]
Javascript system seems more challanging, I like challanges. 😄

**Comment by SebastianStehle** [comment]
We can do both, it is just the question to structure it correctly. If you can support the IQueryProvider in GraphQL and Rest/Swagger we moved a big step forward :)

**Comment by pushrbx** [comment]
Which IQueryProvider? The built-in one, or the one you mentioned above?
Btw I did something, im not sure if it close to what you were explaining earlier:
https://github.com/pushrbx/squidex/commit/f6c87afc1832ba83efc89e3f521c13dc0e1ae5a7
not tested tho

**Comment by pushrbx** [comment]
Cleanup and moved around things a bit: https://github.com/pushrbx/squidex/commit/110df8971694f4d1e3851c2a2ab7d4856364a392
Also I've added a skeleton for supporting GraphQL. My only problem now that I have to come up with an adapter/abstraction which would give back a Schema object for joined schemas. Something which is not being persisted in the database.
Is it enough if I just use `Schema.Create()` method if I want to have a type for a cross-section of two or more Schema objects?

**Comment by SebastianStehle** [comment]
There are a lot of things I would do differently, but lets focus on the GraphQL or Swagger Stuff first. What do you mean with Joined schemas? I though we agree that the query will return results for a specific schema.

**Comment by pushrbx** [comment]
Uh, you are right, in this context there is no such thing as "joined schemas". Is that right if I say the query's result set needs a schema?

**Comment by SebastianStehle** [comment]
No, because the IQueryModule should only return queries that are suitable for the passed in schema.

**Comment by pushrbx** [comment]
Understood. It seems I confused myself with a stray thought.

**Comment by pushrbx** [comment]
![image](https://user-images.githubusercontent.com/6832715/31381587-e6cf2392-adac-11e7-9fd4-42703d94395a.png)

![image](https://user-images.githubusercontent.com/6832715/31381603-f1921bcc-adac-11e7-90f8-9c356ce6e991.png)

Swagger is somewhat done and the controller is working so far. Swagger needs a bit more work to have the right descriptions in there.

Also it comes up in graphql as well:

![image](https://user-images.githubusercontent.com/6832715/31381754-59e570de-adad-11e7-8788-7373575f9c1f.png)
Shouldn't we prefix the name with something?
Now I only need to create test cases, to see if they work properly.

**Comment by SebastianStehle** [comment]
Awesome

I think we should not prefix it or with something nice, e.g. "get"

we also need to test what happens when we create an invalid return value, e.g.  just a number GraphQL and duplicate names and all this stupid stuff.

**Comment by pushrbx** [comment]
How would I create something which would give back an invalid field value for GraphQL?
I've done the following in the plugin:
```cs
public async Task<(ISchemaEntity Schema, long Total, IReadOnlyList<IContentEntity> Items)> Execute(ISchemaEntity schema, QueryContext context, IDictionary<string, object> arguments)
{
    if (arguments == null || arguments.Count == 0 || !arguments.ContainsKey("tagNames"))
    {
        return (schema, 0, new List<IContentEntity>());
    }

    var filterList = new List<string>();
    var parts = arguments["tagNames"].ToString().Split(",");
    foreach (var part in parts)
    {
        filterList.Add($"data/title/en eq '{part}'");
    }

    var filter = string.Join(" or ", filterList);
    var terms = await context.QueryContentsAsync("term", $"$filter={filter}");

    filterList.Clear();

    foreach (var term in terms)
    {
        filterList.Add($"data/terms/iv eq '{term.Id}'");
    }

    filter = string.Join(" or ", filterList);
    var resultContents = await context.QueryContentsAsync(schema.Name, $"$filter={filter}");

    return (schema, resultContents.Count, resultContents);
}
```

Actually what I should ask for is a list of things to test.
What I have tested so far:
- Normal graphql usage -> works
- Normal rest api usage -> works
- Duplicate names -> I use FirstOrDefault, so always the first one is going to be selected.
`query = squidexPlugin.GetQueries(App.Name, schema).FirstOrDefault(x => x.Name == queryName);`

**Comment by SebastianStehle** [comment]
You could query from another schema instead

**Comment by pushrbx** [comment]
This is funny. It's working with different schema. 😄 
```cs
public async Task<(ISchemaEntity Schema, long Total, IReadOnlyList<IContentEntity> Items)> Execute(ISchemaEntity schema, QueryContext context, IDictionary<string, object> arguments)
{
    if (arguments == null || arguments.Count == 0 || !arguments.ContainsKey("tagNames"))
    {
        return (schema, 0, new List<IContentEntity>());
    }

    var filterList = new List<string>();
    var parts = arguments["tagNames"].ToString().Split(",");
    foreach (var part in parts)
    {
        filterList.Add($"data/title/en eq '{part}'");
    }

    var filter = string.Join(" or ", filterList);
    // var terms = await context.QueryContentsAsync("term", $"$filter={filter}");
    var resultContents = await context.QueryContentsAsync("term", $"$filter={filter}");

    //filterList.Clear();

    //foreach (var term in terms)
    //{
    //    filterList.Add($"data/terms/iv eq '{term.Id}'");
    //}

    //filter = string.Join(" or ", filterList);
    //var resultContents = await context.QueryContentsAsync(schema.Name, $"$filter={filter}");

    return (schema, resultContents.Count, resultContents);
}
```
Graphql Request:
```graphql
{
  getNodesWithTag(tagNames: "1998,1999,2000") {
    id
  }
}
```
Response:
```json
{
  "data": {
    "getNodesWithTag": [
      {
        "id": "a7704238-a9a7-4458-b422-c3d958ecaab3"
      },
      {
        "id": "7eba5836-6f8f-4fbc-a908-4503e4adf1d0"
      },
      {
        "id": "5a20a4bb-5246-4ddc-81da-59e0fa143b5f"
      },
      {
        "id": "e765e871-1344-4310-b3fa-1feec0e5fed6"
      }
    ]
  }
}
```

**Comment by pushrbx** [comment]
It works with different schema, but if there is a field which is not in that schema, it returns `null` for it.

**Comment by pushrbx** [comment]
Added PR to review where we are: https://github.com/Squidex/squidex/pull/138

**Comment by SebastianStehle** [comment]
Thats okay, I also could not tell GraphQL to ignore null's.
